clients/drcachesim/tools/opcode_mix.cpp

```diff
@@ -170,10 +170,17 @@ opcode_mix_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
             return false;
         }
         /* If we are dealing with a regdeps trace, we need to set the dcontext ISA mode
-         * to the correct synthetic ISA (i.e., DR_ISA_REGDEPS).
+         * to the correct synthetic ISA (i.e., DR_ISA_REGDEPS). Because isa_mode in
+         * dcontext is a global resource, we guard its access to avoid data races
+         * (even though this is a benign data race, as all threads are writing the same
+         * isa_mode value).
          */
-        if (TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, memref.marker.marker_value))
-            dr_set_isa_mode(dcontext_.dcontext, DR_ISA_REGDEPS, nullptr);
+        if (TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, memref.marker.marker_value)) {
+            std::lock_guard<std::mutex> guard(dcontext_mutex_);
+            dr_isa_mode_t isa_mode = dr_get_isa_mode(dcontext_.dcontext);
```

Comment:
Is it always guaranteed that a shard has a TRACE_MARKER_TYPE_FILETYPE marker before all memref instructions?
Otherwise this won't work and I'll have to guard the whole `decode_from_copy()` routine below (which is where dcontext isa_mode is read).

Reply:
If there is a filetype, it is in the header before any instructions (though maybe that has no invariant check and maybe is not written down...it would be ok to fail with a fatal error there though).  You can use the stream's get_filetype() which caches the filetype once seen if that helps.

Reply:
Ok, failing if we encounter instructions before we have a filetype for now.
Using stream's `get_filetype()` is the right way to do it, but it requires an update of opcode_mix to use `parallel_shard_init_stream()` rather than the now deprecated `parallel_shard_init()`, which does not have a stream.
I think it's outside the scope of this PR (which aims to just fix a data race). I'll add a comment and a new issue about updating opcode_mix.

Reply:
xref #6812.